DCOS-59167 [DS] [Spark Operator] Make metrics reporting configurable and add ServiceMonitor/scraping for the Operator at the deployment step (#17)
mesosphere/kudo-spark-operator: MONITORING.md

> The `kudo-spark-operator` is able to seamlessly integrated with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).

Reviewer: ```suggestion
The `kudo-spark-operator` integrates seamlessly with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).
```

> The `kudo-spark-operator` is able to seamlessly integrated with Prometheus, which is installed within [Prometheus operator](https://github.com/coreos/prometheus-operator).

Reviewer 2: ```suggestion
By default, KUDO Spark Operator supports metrics reporting and integrates with [Prometheus operator](https://github.com/coreos/prometheus-operator).
```

> Integration with other Prometheus distributions, like kube-prometheus, wasn't tested. 

Reviewer 2: ```suggestion
```

> The `prometheus-operator` uses crafted services discovery approach, introducing `ServiceMonitor` kind. 

Reviewer 2: Prometheus Operator relies on `ServiceMonitor` kind which describes the set of targets to be monitored. KUDO Spark Operator configures `ServiceMonitor`s for both the Operator and submitted Spark Applications automatically when monitoring is enabled.

Author: Thanks!

Reviewer 2: didn't work as a suggestion for some reason, do you think we still need the sentence above?

Author: it removed

> But the `kudo-spark-operator` take its configuration burden on itself.

Author: Thanks, Sasha

> #### How to get `kudo-spark-operator` and app's metrics available for scrape by Prometheus:

Reviewer 2: ```suggestion
#### Configuring Spark Operator and Spark Application metrics export to Prometheus
```

1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.

> 1) Install the `kudo-spark-operator` with `-p enableMetrics=true` parameter. (But it's true by default, so typically no need to specify it explicitly)

Reviewer 2: ```suggestion
1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.
```

Author: Thanks, Anton!

> 1) Install the `kudo-spark-operator` with `-p enableMetrics=true` parameter. (But it's true by default, so typically no need to specify it explicitly)

Author: Fixed by Anton's suggestion apply.

1) Composing your Spark Application yaml:

>    - use Spark image with JMXPrometheus exporter jar on the board i.e. `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus` 

Author: I agree, here should be `mesosphere/spark` image. The reason why I didn't put it here this time - the aforementioned image isn't verified in terms of metrics exporting. I remember you have done good job adding Prometheus configs there, but the image should be assembled, pushed into Docker Hub and checked that the metrics exported. I'm going to do in scope of other PR (introducing Operator dashboard). So, I'd like leave it as is here at this time.

>    - use Spark image with JMXPrometheus exporter jar on the board i.e. `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus` 

Reviewer: > the image should be assembled, pushed into Docker Hub and checked that the metrics exported

yes, this is basically what I've done in scope of testing #12 to verify the metrics are exported into Prometheus. The only difference is that I used my personal docker hub to do that. Since it's a readme file, I don't see any issue with specifying `mesosphere/image` here at the first place just to prevent any confusion or just to not to forget to do it later

>    - use Spark image with JMXPrometheus exporter jar on the board i.e. `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus` 

Reviewer: You can use this image `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s`

>    - use Spark image with JMXPrometheus exporter jar on the board i.e. `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus` 

Author: Thanks @alembiewski. As we discussed, metrics exporting is verified with this image. So it makes sense to mention it here instead of Google's one.

>    - enable driver/executors monitoring by adding the yaml piece into `spec` section:

Reviewer 2: ```suggestion
   - enable Driver and Executors metrics reporting by adding the following configuration into `SparkApplication` `spec` section:
```

     ```yaml
       monitoring:
         exposeDriverMetrics: true
         exposeExecutorMetrics: true
         prometheus:
           jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.11.0.jar"
           port: 8090
     ```  

>    - if you would like use other than 8090 port for metrics exporting, you must pass appropriate parameter during `kudo-spark-operator` installation `-p appMetricsPort=<desired_port>` 

Reviewer 2: ```suggestion
   - if you would like to use a different port for metrics exporting, it should be provided via `-p appMetricsPort=<desired_port>` during KUDO Spark Operator installation
```

>    - Mark your application by the label 

Reviewer: ```suggestion
   - Mark your application with the label 
```

>    - Mark your application by the label 

Reviewer 2: ```suggestion
   - specify additional labels for the application to make it discoverable by `ServiceMonitor`
```

     ```yaml
       labels:
         exposedMetrics: true
     ```

> Full blown example you can find in [spark-application.yaml](specs/spark-application.yaml).

Reviewer 2: ```suggestion
Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).
```

Author: Could you, pls, shine light on why the wording better?

Reviewer 2: Sure, here are some points:
* original sentence lacks the subject (example of what?)
* word ordering issue (should be 'you can find the example')
* we avoid using pronouns ('you') in the service docs
* 'Full blown' doesn't suit the context here (see [Merriam-Webster](https://www.merriam-webster.com/dictionary/full-blown) definition and synonyms)
* `spark-application.yaml` -> `specs/spark-application.yaml` in the link name provides the location of the file in the project,  not just the link

Does it make sense?

Author: @akirillov, it's sweet :-) Thanks for the explanation. It makes sense. Just a couple of comments from my side:  
> original sentence lacks the subject (example of what?)

neat!

> we avoid using pronouns ('you') in the service docs

Good point. And If you have somewhere guide/list of best practices like that, which we should follow. I'd appreciate if you will share it with entire team.

> 'Full blown' doesn't suit the context here (see Merriam-Webster definition and synonyms)

I do not insist to use `Full blown' phrase here :-) just to debate - as per [Collins dictionary](https://www.collinsdictionary.com/dictionary/english/full-blown) it also has meaning "fully grown or developed;", which is suitable for this context, imo.

> spark-application.yaml -> specs/spark-application.yaml in the link name provides the location of the file in the project, not just the link

mb it makes sense for better transparency. Beside of that user always can see location in url or, after click, in github ui.

Reviewer 2: you can check other services' docs created by the team for the inspiration regarding the style and format e.g. [Data Science Engine](https://docs.d2iq.com/mesosphere/dcos/services/data-science-engine/1.0.0/) or [DCOS Spark](https://docs.d2iq.com/mesosphere/dcos/services/spark/2.9.0-2.4.3/). I think Docs team has some guidelines but usually, we follow review suggestions.

RE: 'full blown', I don't think it's wrong it's just doesn't suit the context: 'fully grown configuration example' or 'fully developed configuration example' (developed not in the software sense). I'd expect 'full' or 'complete' here, but it's up to you.

Author: ok, thanks, make sense

Out of the box, the `kudo-spark-operator` has enabled metrics reporting. 

> By default, it supports integration with [Prometheus operator](https://github.com/coreos/prometheus-operator).

Reviewer 3: ```suggestion
By default, it supports integration with the [Prometheus operator](https://github.com/coreos/prometheus-operator).
```

The `prometheus-operator` uses crafted services discovery approach, introducing `ServiceMonitor` kind. 
Prometheus Operator relies on `ServiceMonitor` kind which describes the set of targets to be monitored. 
KUDO Spark Operator configures `ServiceMonitor`s for both the Operator and submitted Spark Applications automatically 
when monitoring is enabled.

> #### Configuring Spark Operator and Spark Application metrics export to Prometheus

Reviewer 3: ```suggestion
#### Exporting Spark Operator and Spark Application metrics to Prometheus
```

1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.
1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.
1) Create ServiceMonitor for Spark (see prometheus-operator docs). Take this yaml without modification - 
   ```yaml
   cat <<EOF | kubectl apply -f -
   apiVersion: monitoring.coreos.com/v1

>     kind: ServiceMonitor

Reviewer 3: ```suggestion
   kind: ServiceMonitor
```

   metadata:
     labels:
       app: prometheus-operator
       release: prometheus-kubeaddons
     name: spark-cluster-monitor
   spec:
     endpoints:
       - interval: 5s
         port: metrics
     selector:
       matchLabels:
         spark/servicemonitor: "true"
   EOF
   ```
1) Composing your Spark Application yaml:

>    - use Spark image with JMXPrometheus exporter jar on the board i.e. `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s` 

Reviewer 3: ```suggestion
   - use the following Spark image which includes the `JMXPrometheus` exporter jar: `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s` 
```

   - enable driver/executors monitoring by adding the yaml piece into `spec` section:
     ```yaml
       monitoring:
         exposeDriverMetrics: true
         exposeExecutorMetrics: true
         prometheus:
           jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.11.0.jar"
           port: 8090
     ```  
   - if it's necessary to expose metrics endpoint on other than 8090 port, then, please, 

>      1) change the port value in `SparkApplication` definition yaml (`spec.monitoring.prometheus.port`) and 

Reviewer 3: ```suggestion
     1) change the `port` value in the `SparkApplication` yaml definition (`spec.monitoring.prometheus.port`)
```

>      1) specify appropriate parameter during `kudo-spark-operator` installation - `-p appMetricsPort=<desired_port>`. 

Reviewer 3: Can you share a fuller cmd example? Something like:

```suggestion
     1) specify the same port when installing the `kudo-spark-operator`:  
     ```
     kubectl kudo install <operator> -p appMetricsPort=<desired_port>
     ```
```

   - Mark `driver` and/or `executor` with the label `metrics-exposed: "true"` -
     ```yaml
     spec:
       driver:
         labels:
            metrics-exposed: "true"
       executor:
         labels:
           metrics-exposed: "true"
     ```

> Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).

Reviewer 3: The image in this example uses `gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus`. Should we update it so that the instructions match?

Reviewer 3: Also, this example file does not have the `metrics-exposed` labels as suggested above. Might confuse users.

Author: > Also, this example file does not have the metrics-exposed labels as suggested above. Might confuse users.
It has :-) Take a look, pls, [the label in executor spec](https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-94f8071a156ee28e0f01545179230631R37) and [the one in driver spec](https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-94f8071a156ee28e0f01545179230631R26)

Author: > The image in this example uses gcr.io/spark-operator/spark:v2.4.4-gcs-prometheus. Should we update it so that the instructions match?

Done. https://github.com/mesosphere/kudo-spark-operator/pull/17/files#diff-94f8071a156ee28e0f01545179230631R8

Reviewer 3: So when I went through the instructions I got up to this point and then thought, "what next?" 

At this point you've showed users can install the operator and set the configs needed in the spark application to enable metrics. But the docs stop before you instruct users:

- to deploy the spark application
- find the URL for the prometheus dashboard
- or if there a `/metrics` endpoint that users can go to see which metrics are being sent?

The final conclusion of this doc should be something like 
```
Now you can view metrics by going to <url> and seeing the spark metrics being sent
```
or 
```
Now go to the prometheus dashboard at <url> and search for metrics starting with 'spark'
```
Let's be empathetic to the end users

Author: @samvantran very good point, thanks for catching it! 
Addressed.

Out of the box, the `kudo-spark-operator` has enabled metrics reporting. 
By default, it supports integration with the [Prometheus operator](https://github.com/coreos/prometheus-operator).

Prometheus Operator relies on `ServiceMonitor` kind which describes the set of targets to be monitored. 
KUDO Spark Operator configures `ServiceMonitor`s for both the Operator and submitted Spark Applications automatically 
when monitoring is enabled.

#### Exporting Spark Operator and Spark Application metrics to Prometheus

> 1) Ensure `prometheus-operator` is installed on your Kubernetes cluster.

Reviewer 3: Can you provide a link that shows users what instructions are needed to install the prometheus-operator? And some information on how they can get to the prometheus dashboard once installed? 

Or are you expecting users to be able to access the pre-installed prom/grafana charts that come with konvoy?

Author: Yes, I'm expecting users install the operator to Konvoy, where all of these tools (Prometheus, Grafana) are installed.
On the other hand, we agreed to make the kudo-spark-operator able to run on any kubernetes cluster. So, it make sense provide link to an instruction how to install the prometheus-operator like [this](https://github.com/coreos/prometheus-operator#quickstart).  My only concern I don't personally tried install `prometheus-operator` on vanilla kubernetes cluster and maybe can't help a user in case of any troubles with it.

Author: addressed

> 1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.

Author: I'm expecting the user will come to the guide from README.md, where he followed instruction how to install the operator. I think we can remove this point from the instruction. I will introduce prerequisite section, where the kudo-spark-operator will be mentioned

> 1) Install the KUDO Spark Operator. Metrics reporting is enabled by default and can be disabled by modifying `enableMetrics` parameter.

Author: Addressed

1) Create ServiceMonitor for Spark (see prometheus-operator docs). Take this yaml without modification - 

>    ```yaml

Author: fixed

   cat <<EOF | kubectl apply -f -
   apiVersion: monitoring.coreos.com/v1
   kind: ServiceMonitor
   metadata:
     labels:
       app: prometheus-operator
       release: prometheus-kubeaddons
     name: spark-cluster-monitor
   spec:
     endpoints:
       - interval: 5s
         port: metrics
     selector:
       matchLabels:
         spark/servicemonitor: "true"
   EOF
   ```
1) Composing your Spark Application yaml:
   - use the following Spark image which includes the `JMXPrometheus` exporter jar: `mesosphere/spark:2.4.4-bin-hadoop2.7-k8s` 
   - enable Driver and Executors metrics reporting by adding the following configuration into `SparkApplication` `spec` section:
     ```yaml
       monitoring:
         exposeDriverMetrics: true
         exposeExecutorMetrics: true
         prometheus:
           jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.11.0.jar"
           port: 8090
     ```  

>    - if it's necessary to expose metrics endpoint on other than 8090 port, then, please, 

Reviewer 3: ```suggestion
   - if it's necessary to expose the metrics endpoint on a port other than `8090`, do the following:
```

Out of the box, the `kudo-spark-operator` has enabled metrics reporting. 
By default, it supports integration with the [Prometheus operator](https://github.com/coreos/prometheus-operator).

Prometheus Operator relies on `ServiceMonitor` kind which describes the set of targets to be monitored. 
KUDO Spark Operator configures `ServiceMonitor`s for both the Operator and submitted Spark Applications automatically 
when monitoring is enabled.

#### Exporting Spark Operator and Spark Application metrics to Prometheus

##### Prerequisite
* The *`prometheus-operator`*.

> If you are using Kubernetes cluster without pre-installed the `prometheus-operator` then follow 

Reviewer 3: ```suggestion
If you don't already have the `prometheus-operator` installed on your Kubernetes cluster, you can do so by following
```

   - enable Driver and Executors metrics reporting by adding the following configuration into `SparkApplication` `spec` section:
     ```yaml
       monitoring:
         exposeDriverMetrics: true
         exposeExecutorMetrics: true
         prometheus:
           jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.11.0.jar"
           port: 8090
     ```  
   - if it's necessary to expose the metrics endpoint on a port other than `8090`, do the following:
     1) change the `port` value in the `SparkApplication` yaml definition (`spec.monitoring.prometheus.port`)
     1) specify the same port when installing the `kudo-spark-operator`:  
     ```
     kubectl kudo install <operator> -p appMetricsPort=<desired_port>
     ```
   - Mark `driver` and/or `executor` with the label `metrics-exposed: "true"` -
     ```yaml
     spec:
       driver:
         labels:
            metrics-exposed: "true"
       executor:
         labels:
           metrics-exposed: "true"
     ```
   - Install the SparkApplication:
   ```
   kubectl apply -f <path_to_the_application_yaml>   
   ```

>    Full configuration example is available in [specs/spark-application.yaml](specs/spark-application.yaml).

Reviewer 3: All of these should be indented two spaces so that they are aligned properly with `- Install the SparkApplication`

> 1) Create the metrics endpoint service. And don't forget to modify the service port in the yaml in case you have changed it 

Reviewer 3: Shouldn't we set up the metrics endpoint service before submitting the spark application? Feels like this step should go right after 
```
1) Create a `ServiceMonitor` for Spark: 
```

Author: I placed those steps in that order just to simplify the port changing steps sequence. I tried swap them, PTAL.

Reviewer 3: 👍

on previous step.
   ```bash
   cat <<EOF | kubectl apply -f - 
   apiVersion: v1
   kind: Service
   metadata:
     name: spark-application-metrics
     labels:
       "spark/servicemonitor": "true"
   spec:
     ports:
       - port: 8090
         name: metrics
     clusterIP: None
     selector:

>        "metrics-exposed": "true"

Reviewer 3: missing `EOF`

   ```  

> 1) Now go to the prometheus dashboard at `<kubernetes_endpoint_url>/ops/portal/prometheus/graph` and search for metrics 

Reviewer 3: ```suggestion
1) Now, go to the prometheus dashboard (e.g. `<kubernetes_endpoint_url>/ops/portal/prometheus/graph`) and search for metrics 
```

> starting with 'spark'. The Prometheus uri might be differ depends on the `prometheus-operator` installation configuration. 

Reviewer 3: ```suggestion
starting with 'spark'. The Prometheus URI might be different depending on how you configured and installed the `prometheus-operator`. 
```
